### PR TITLE
Implement filter.

### PR DESCRIPTION
**Pull request by spacemeowx2**
- [x] Test on Linux with libpcap
- [ ] Test on Linux with pfring
- [x] Test on Windows with WinPcap or npcap
- [x] Test on macOS

### CODE REVIEW

**Comment by spacemeowx2** [comment]
I can't make pfring work.

```
sudo ./target/debug/examples/filter
Opening packet capturing library
Library opened, version is pfring 7.5.0
Opening the lo interface
Interface opened, data link: ethernet
./target/debug/examples/filter: symbol lookup error: /usr/local/lib/libpfring.so.1: undefined symbol: pcap_open_dead
```


Seems that `libpfring` doesn't dynamic link with `libpcap`.

```
ldd -r /usr/local/lib/libpfring.so.1
        linux-vdso.so.1 (0x00007ffeca397000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fd96b223000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fd96b01b000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fd96ae17000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fd96aa26000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fd96b6de000)
undefined symbol: pcap_compile_nopcap   (/usr/local/lib/libpfring.so.1)
undefined symbol: pcap_close    (/usr/local/lib/libpfring.so.1)
undefined symbol: pcap_open_dead        (/usr/local/lib/libpfring.so.1)
undefined symbol: pcap_compile  (/usr/local/lib/libpfring.so.1)
undefined symbol: pcap_freecode (/usr/local/lib/libpfring.so.1)
undefined symbol: bpf_filter    (/usr/local/lib/libpfring.so.1)
```

---

Finally I find out I can make it work with this command:

`LD_PRELOAD=/usr/local/lib/libpcap.so.1.8.1 ./target/debug/examples/filter`

I think we need to load `libpcap` before `set_filter`. How can we load `libpcap` by code?

**Comment by szymonwieloch** [comment]
@spacemeowx2 Hi! Thank you for your contribution. I am dead tired today because I had 2 hours of training. I will review your changes tomorrow, when I have more time. Cheers!

File: src/traits.rs [hunk 1]
```diff
@@ -37,9 +37,10 @@ pub trait DynamicInterface<'a>: Send + Sync{
     */
     fn loop_infinite_dyn(&self, callback: & dyn FnMut(&BorrowedPacket)) -> Result<(), Error>;
 
+    ///Set bpf filter.
+    fn set_filter(&mut self, filter: &str) -> Result<(), Error>;
```
Comment (szymonwieloch) [inline src/traits.rs]: Passing&str as an agrument has a drawback - it requires memory allocation for conversion to CString. Most of the time this allocation is not necessary. A better option would be to create two functions: set_filter() and set_filter_cstr(filter: &CStr) - this gives user a choice.

set_filter() can then be implemented in the trait (traits allow default implementations) only as a conversion function:

```
fn set_filter(&mut self, filter: &str) -> Result<(), Error> {
    let filter = CString::new(filter)?;
    self.set_filter_cstr(&filter)
}   
```

File: src/wpcap/interface.rs [hunk 2]
```diff
@@ -16,7 +17,8 @@ pub struct Interface<'a> {
     handle: * const PCapHandle,
     dll: & 'a WPCapDll,
     datalink: DataLink,
-    queue: * mut PCapSendQueue
+    queue: * mut PCapSendQueue,
+    bpf_filter: Option<BpfProgram>,
```
Comment (szymonwieloch) [inline src/wpcap/interface.rs]: It is not necessary to keep the BpfProgram instance inside the Interface. Please check the documentation:

https://www.tcpdump.org/manpages/pcap_freecode.3pcap.html

You can call pcap_freecode just after calling pcap_setfilter.

File: src/traits.rs [hunk 3]
```diff
@@ -37,9 +37,10 @@ pub trait DynamicInterface<'a>: Send + Sync{
...
+
```
Comment (szymonwieloch) [inline src/traits.rs:52]: This API clearly lacks a function that would remove a filter:

fn remove_filter(&self);

PF_RING has related function: pffring_remove_bpf_filer(). Pcap does not have similar function but my guess is that settig empty filter has the same effect:
```
self.set_filter_cstr(const_cstr!(""))
```

File: src/wpcap/interface.rs [hunk 4]
```diff
@@ -49,21 +51,29 @@ impl<'a> Interface<'a> {
...
+            bpf_filter: None,
         })
     }
 
     fn last_error(&self) -> Error {
         let cerr = unsafe{self.dll.pcap_geterr(self.handle)};
         Error::LibraryError(cstr_to_string(cerr))
     }
+
+    fn drop_filter(&mut self) {
```
Comment (szymonwieloch) [inline src/wpcap/interface.rs]: This function is not necessary - once you call pcap_setfiler() the filter is applied and the structure is no longer used. pcap_freecode only releases resources related to the BpfProgram. There is no effect from the user perspective.

File: src/wpcap/interface.rs [hunk 5]
```diff
@@ -144,6 +154,23 @@ impl<'a> traits::DynamicInterface<'a> for Interface<'a> {
             Err(self.last_error())
         }
     }
+
+    fn set_filter(&mut self, filter: &str) -> Result<(), Error> {
+        let mut bpf_filter: BpfProgram = unsafe {uninitialized()};
+        let filter = CString::new(filter)?;
+        let result = unsafe { self.dll.pcap_compile(self.handle, &mut bpf_filter, filter.as_ptr(), 1, 0) };
```
Comment (szymonwieloch) [inline src/wpcap/interface.rs:151]: According to the documentation, prior, to pcap 1.8 this function was not thread safe:

https://www.tcpdump.org/manpages/pcap_compile.3pcap.html

So we need a mutex to make the library work with older versions:

```
lazy_static! {
	static ref COMPILE_GUARD : Mutex<()> = Mutex::new();
}
```

and then:
```
let result = {
    let _lock = COMPILE_GUARD.lock().unwrap();
    unsafe { self.dll.pcap_compile(self.handle, &mut bpf_filter, filter.as_ptr(), 1, 0) }
}
```

File: examples/filter.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    let mut interf = lib.open_interface(&interf_name).expect("Could not open network interface");
+    if let DataLink::Ethernet = interf.data_link() {
+        println!("Interface opened, data link: {}", interf.data_link());
+    } else {
+        println!("data link should be Ethernet, {} found", interf.data_link());
+        return
+    }
+
+    interf.set_filter("ether dst 45:00:00:54:ee:96").expect("Could not set filter");
+    //send some packets
```
Comment (szymonwieloch) [inline examples/filter.rs]: Generally examples are supposed to be as simple as possible. My understanding is that filtering packets is related only to receiving packets, so showing how to send packets in a filtering example may be misleading to some users.

**Review by szymonwieloch — CHANGES_REQUESTED** [review]
Hi! This is a great feature and great code. I have found some issues with the implementation that require fixing before merging. Please let me know if you want to fix them by yourself or if you want me to do it.

**Comment by szymonwieloch** [comment]
@spacemeowx2 When it comes to loading a pcap library, I think it should be delayed as long as possible. Many users may not need to perform filtering and it is pointless to use to much resources when not needed. You are right - the best place to do it is probably in the set_filter() function. My guess would be that you need to add it to the Library instance:

```
pcap: Mutex<Option<dlopen::raw::Library>>
```
and add a prefetch_pcap() function : 
```
pub (crate) fn prefetch_pcap(&self){}
```

Then call prefetch_pcap() from set_filter()

I haven't tested it though and I am curious if it is possible to load pcap library in this way...

**Comment by spacemeowx2** [comment]
Ok, I'll fix it by myself.

**Comment by spacemeowx2** [comment]
@szymonwieloch After many attempts, I found that calling `dlopen` with `RTLD_GLOBAL | RTLD_LAZY` or `RTLD_LOCAL | RTLD_NOW` allows `pcap` to be loaded and used by `pfring`. However, `dlopen` crate uses the flag `RTLD_LOCAL | RTLD_LAZY` and cannot be changed at runtime.

I guess I will leave this problem to you to solve.

**Comment by tan-wei** [comment]
It seems that `set_filter` not support in version `0.3.0`, maybe the new interface is not merged in the release crate?

**Pull request merged**